File: packages/ttsc-plugin/native/host/manifest.go
package host

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"

	shimast "github.com/microsoft/typescript-go/shim/ast"

	"github.com/ts-refinement/ttsc-plugin/native/analysis"
)

const (
	refinementManifestFileName      = ".ts-refinement-manifest.json"
	refinementManifestSchemaVersion = 1
	refinementMarkerPrefix          = "ts-refinement-site:"
)

type nativeManifestAsset struct {
	File   string `json:"file"`
	SHA256 string `json:"sha256"`
}

type nativeManifestSite struct {
	ID            string   `json:"id"`
	Length        int      `json:"length"`
	Module        string   `json:"module"`
	PredicateKeys []string `json:"predicateKeys"`
	Start         int      `json:"start"`
}

type nativeManifestCheckIdentity struct {
	Path          []analysis.PathSegment `json:"path"`
	PredicateKeys []string               `json:"predicateKeys"`
}

type nativeManifest struct {
	Assets        []nativeManifestAsset `json:"assets"`
	BuildID       string                `json:"buildId"`
	Project       nativeManifestProject `json:"project"`
	SchemaVersion int                   `json:"schemaVersion"`
	Sites         []nativeManifestSite  `json:"sites"`
}

type nativeManifestProject struct {
	ConfigPath string `json:"configPath"`
}

type nativeBuildTracker struct {
	buildID    string
	configPath string
	sites      map[string]nativeManifestSite
}

func newNativeBuildTracker(cwd, tsconfig string) (*nativeBuildTracker, error) {
	buildID, err := randomUUID()
	if err != nil {
		return nil, err
	}
	configPath := tsconfig
	if !filepath.IsAbs(configPath) {
		configPath = filepath.Join(cwd, configPath)
	}
	return &nativeBuildTracker{
		buildID:    buildID,
		configPath: filepath.Clean(configPath),
		sites:      map[string]nativeManifestSite{},
	}, nil
}

func (tracker *nativeBuildTracker) register(
	file *shimast.SourceFile,
	site assertion,
	checks []analysis.Check,
	recursions []analysis.Recursion,
) string {
	start := tokenStart(file, site.node)
	predicateKeys := []string{}
	checkIdentities := make([]nativeManifestCheckIdentity, 0, len(checks))
	for _, check := range checks {
		checkPredicateKeys := make([]string, 0, len(check.Definition.Predicates))
		for _, predicate := range check.Definition.Predicates {
			key := predicate.Key()
			predicateKeys = append(predicateKeys, key)
			checkPredicateKeys = append(checkPredicateKeys, key)
		}
		checkIdentities = append(checkIdentities, nativeManifestCheckIdentity{
			Path:          check.Path,
			PredicateKeys: checkPredicateKeys,
		})
	}
	module, err := filepath.Rel(filepath.Dir(tracker.configPath), file.FileName())
	if err != nil {
		module = file.FileName()
	}
	identity := struct {
		Checks        []nativeManifestCheckIdentity `json:"checks"`
		Length        int                           `json:"length"`
		Module        string                        `json:"module"`
		PredicateKeys []string                      `json:"predicateKeys"`
		Recursions    []analysis.Recursion          `json:"recursions"`
		Start         int                           `json:"start"`
	}{
		Checks:        checkIdentities,
		Length:        site.node.End() - start,
		Module:        filepath.ToSlash(module),
		PredicateKeys: predicateKeys,
		Recursions:    recursions,
		Start:         start,
	}
	encoded, _ := json.Marshal(identity)
	idBytes := sha256.Sum256(encoded)
	id := hex.EncodeToString(idBytes[:])
	tracker.sites[id] = nativeManifestSite{
		ID:            id,
		Length:        identity.Length,
		Module:        identity.Module,
		PredicateKeys: predicateKeys,
		Start:         start,
	}
	return refinementMarkerPrefix + tracker.buildID + ":" + id
}

func (tracker *nativeBuildTracker) write(
	directory string,
	assets []nativeManifestAsset,
) error {
	sites := make([]nativeManifestSite, 0, len(tracker.sites))
	for _, site := range tracker.sites {
		sites = append(sites, site)
	}
	sort.Slice(sites, func(left, right int) bool {
		if sites[left].Module == sites[right].Module {
			return sites[left].Start < sites[right].Start
		}
		return sites[left].Module < sites[right].Module
	})
	sort.Slice(assets, func(left, right int) bool {
		return assets[left].File < assets[right].File
	})
	manifest := nativeManifest{
		Assets:        assets,
		BuildID:       tracker.buildID,
		Project:       nativeManifestProject{ConfigPath: tracker.configPath},
		SchemaVersion: refinementManifestSchemaVersion,
		Sites:         sites,
	}
	encoded, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return err
	}
	encoded = append(encoded, '\n')
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(directory, refinementManifestFileName), encoded, 0o644)
}

func nativeManifestAssetFor(directory, fileName, source string) (nativeManifestAsset, bool) {
	switch filepath.Ext(fileName) {
	case ".cjs", ".js", ".mjs":
	default:
		return nativeManifestAsset{}, false
	}
	relative, err := filepath.Rel(directory, fileName)
	if err != nil {
		return nativeManifestAsset{}, false
	}
	digest := sha256.Sum256([]byte(source))
	return nativeManifestAsset{
		File:   filepath.ToSlash(relative),
		SHA256: hex.EncodeToString(digest[:]),
	}, true
}

func randomUUID() (string, error) {
	value := make([]byte, 16)
	if _, err := rand.Read(value); err != nil {
		return "", fmt.Errorf("generate refinement build ID: %w", err)
	}
	value[6] = (value[6] & 0x0f) | 0x40
	value[8] = (value[8] & 0x3f) | 0x80
	return fmt.Sprintf(
		"%s-%s-%s-%s-%s",
		hex.EncodeToString(value[0:4]),
		hex.EncodeToString(value[4:6]),
		hex.EncodeToString(value[6:8]),
		hex.EncodeToString(value[8:10]),
		hex.EncodeToString(value[10:16]),
	), nil
}

File: packages/unplugin/src/manifest.ts
import { createHash, randomUUID } from "node:crypto";
import { mkdirSync, writeFileSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, extname, relative, resolve } from "node:path";

import {
  refinementManifestSchemaVersion,
  refinementManifestFileName,
  refinementSiteMarker,
  type AnalysisResult,
  type RefinementManifest,
  type RefinementManifestAsset,
  type RefinementManifestSite,
} from "@ts-refinement/analyzer";

export interface BuildTracker {
  readonly buildId: string;
  readonly configPath: string;
  readonly sites: readonly RefinementManifestSite[];
  registerSite(analysis: AnalysisResult): string;
  reset(configPath: string): void;
}

interface OutputBundleEntry {
  readonly code?: string;
  readonly fileName: string;
  readonly source?: string | Uint8Array;
  readonly type: string;
}

interface OutputLocation {
  readonly dir?: string;
  readonly file?: string;
}

export interface FinalJavaScriptAsset {
  readonly file: string;
  readonly source: string | Uint8Array;
}

function normalizeModule(configPath: string, fileName: string): string {
  return relative(dirname(configPath), fileName).replaceAll("\\", "/");
}

function siteIdentity(configPath: string, analysis: AnalysisResult): RefinementManifestSite {
  if (analysis.site.checks.length === 0) {
    throw new Error("Cannot register an unresolved refinement site.");
  }
  const identity = {
    length: analysis.site.node.getWidth(),
    module: normalizeModule(configPath, analysis.site.fileName),
    predicateKeys: analysis.site.checks.flatMap((check) =>
      check.definition.predicates.map((predicate) => predicate.key),
    ),
    start: analysis.site.node.getStart(),
  };
  const id = createHash("sha256")
    .update(
      JSON.stringify({
        ...identity,
        checks: analysis.site.checks.map((check) => ({
          path: check.path,
          predicateKeys: check.definition.predicates.map((predicate) => predicate.key),
        })),
        recursions: analysis.site.recursions,
      }),
    )
    .digest("hex");
  return { id, ...identity };
}

export function createBuildTracker(): BuildTracker {
  let buildId = "";
  let configPath = "";
  const byId = new Map<string, RefinementManifestSite>();

  return {
    get buildId() {
      return buildId;
    },
    get configPath() {
      return configPath;
    },
    get sites() {
      return [...byId.values()].sort((left, right) =>
        left.module === right.module
          ? left.start - right.start
          : left.module.localeCompare(right.module),
      );
    },
    registerSite(analysis) {
      if (buildId.length === 0 || configPath.length === 0) {
        throw new Error("Refinement build tracker was not initialized.");
      }
      const site = siteIdentity(configPath, analysis);
      const existing = byId.get(site.id);
      if (existing !== undefined && JSON.stringify(existing) !== JSON.stringify(site)) {
        throw new Error(`Refinement runtime site ID collision for '${site.id}'.`);
      }
      byId.set(site.id, site);
      return refinementSiteMarker(buildId, site.id);
    },
    reset(nextConfigPath) {
      buildId = randomUUID();
      configPath = nextConfigPath;
      byId.clear();
    },
  };
}

function optionalOutputDirectory(options: OutputLocation): string | null {
  if (options.dir !== undefined) return resolve(options.dir);
  if (options.file !== undefined) return dirname(resolve(options.file));
  return null;
}

function outputDirectory(options: OutputLocation): string {
  const directory = optionalOutputDirectory(options);
  if (directory === null) {
    throw new Error("Unable to determine build output directory for refinement manifest.");
  }
  return directory;
}

function isJavaScriptFile(fileName: string): boolean {
  return [".cjs", ".js", ".mjs"].includes(extname(fileName));
}

function manifestAssets(assets: readonly FinalJavaScriptAsset[]): RefinementManifestAsset[] {
  return assets
    .map((asset) => ({
      file: asset.file.replaceAll("\\", "/"),
      sha256: createHash("sha256").update(asset.source).digest("hex"),
    }))
    .sort((left, right) => left.file.localeCompare(right.file));
}

export function buildManifestSource(
  tracker: BuildTracker,
  assets: readonly FinalJavaScriptAsset[],
): string {
  if (tracker.buildId.length === 0 || tracker.configPath.length === 0) {
    throw new Error("Refinement build tracker was not initialized.");
  }
  const manifest: RefinementManifest = {
    assets: manifestAssets(assets),
    buildId: tracker.buildId,
    project: { configPath: tracker.configPath },
    schemaVersion: refinementManifestSchemaVersion,
    sites: tracker.sites,
  };
  return `${JSON.stringify(manifest, null, 2)}\n`;
}

export async function writeFinalAssetManifest(
  tracker: BuildTracker,
  directory: string,
  assets: readonly FinalJavaScriptAsset[],
): Promise<void> {
  await mkdir(directory, { recursive: true });
  await writeFile(
    resolve(directory, refinementManifestFileName),
    buildManifestSource(tracker, assets),
  );
}

export function writeFinalAssetManifestSync(
  tracker: BuildTracker,
  directory: string,
  assets: readonly FinalJavaScriptAsset[],
): void {
  mkdirSync(directory, { recursive: true });
  writeFileSync(
    resolve(directory, refinementManifestFileName),
    buildManifestSource(tracker, assets),
  );
}

export async function finalAssetsFromPaths(
  directory: string,
  paths: readonly string[],
): Promise<readonly FinalJavaScriptAsset[]> {
  return Promise.all(
    paths.map(async (fileName) => ({
      file: relative(directory, fileName).replaceAll("\\", "/"),
      source: await readFile(fileName),
    })),
  );
}

export async function writeBuildManifest(
  tracker: BuildTracker,
  options: OutputLocation,
  bundle: Readonly<Record<string, OutputBundleEntry>>,
): Promise<void> {
  const directory = outputDirectory(options);
  await writeFinalAssetManifest(
    tracker,
    directory,
    Object.values(bundle).flatMap((entry): readonly FinalJavaScriptAsset[] => {
      if (
        entry.type === "chunk" &&
        entry.code !== undefined &&
        !/\.d\.[cm]?ts$/u.test(entry.fileName)
      ) {
        return [{ file: entry.fileName, source: entry.code }];
      }
      if (
        entry.type === "asset" &&
        entry.source !== undefined &&
        isJavaScriptFile(entry.fileName)
      ) {
        return [{ file: entry.fileName, source: entry.source }];
      }
      return [];
    }),
  );
}
